dev/com.ibm.ws.transport.http/src/com/ibm/ws/genericbnf/internal/BNFHeadersImpl.java

```diff
@@ -848,6 +855,9 @@ public HeaderField getHeader(HeaderKeys key) {
         if (null == key) {
             throw new IllegalArgumentException("Null input provided");
         }
+        if (headerKeysByName.get(key.getName()) == null) {
+            return NULL_HEADER;
+        }
         HeaderElement elem = findHeader(key);
```

Comment:
With this update, we lose the debug statement.  Maybe this should be updated so that we keep the debug statement?

```suggestion
        HeaderElement elem = headerKeysByName.get(key.getName()) == null ? null : findHeader(key);
```

Reply:
Looking at all of this more, I wonder if this all is the right place to do the update and if findHeader should be what is updated instead of each place that calls it.  You did not update containsHeader for instance to do this updated logic that is done in getHeader.  If findHeader was updated to use `headerKeysByName` and return null, wouldn't we have all the same benefits and less maintenance in the future?

Reply:
The other concern I have is how we do ignore case on some of this logic.  Definitely there needs to be some testing of this logic to make sure it works before and after and that we don't have any failures when dealing with keys with different cases.

Reply:
> With this update, we lose the debug statement. Maybe this should be updated so that we keep the debug statement?

The point of this part of the change is to "fail fast" in the case where there is no such header in this particular request or response ... so we don't want to call `findHeader(key)` because we already know it isn't there.

I thought the trace was only useful in the case where the header was present, since the main point of the trace is to produce the header associated with the key. But I can restructure it like this, to avoid the call to `findHeader(key)` and still produce the trace.

```
    public HeaderField getHeader(HeaderKeys key) {
        if (null == key) {
            throw new IllegalArgumentException("Null input provided");
        }
        HeaderElement elem = null;
        if (headerKeysByName.get(key.getName()) != null) {
            elem = findHeader(key);
        }
        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
            Tr.debug(tc, "getHeader(h): " + key.getName() + " " + elem);
        }
        if (null == elem) {
            return NULL_HEADER;
        }
```

Reply:
> Looking at all of this more, I wonder if this all is the right place to do the update and if findHeader should be what is updated instead of each place that calls it. You did not update containsHeader for instance to do this updated logic that is done in getHeader. If findHeader was updated to use `headerKeysByName` and return null, wouldn't we have all the same benefits and less maintenance in the future?

The original main point of this set of changes was to eliminate the nested call to `findKey(header)` in `getHeader(String header)`  which is the next diff below. We can't fix that problem in `findHeader`

The main perf issue here is that finding the key from a String header name is really slow, and most callers to getHeader use the string representation, even if they have the key (which seems odd). To avoid getting the key from the String header the way it is implemented further down the callstack, without mucking around in really complicated old code that I might break, I preferred to use the cache at this level.

Reply:
> The other concern I have is how we do ignore case on some of this logic. Definitely there needs to be some testing of this logic to make sure it works before and after and that we don't have any failures when dealing with keys with different cases.

Very good point ... I'll need to look at this.

Reply:
re: String case

My initial reaction is that the cache should be case-insensitive for the key Strings, so to achieve that I have added `.toLowerCase()` when putting entries in the cache and when checking for them on `getHeader(String name)` calls. 

I'll look for more input on this from a reviewer from channels or web container.

Reply:
> > With this update, we lose the debug statement. Maybe this should be updated so that we keep the debug statement?
> 
> The point of this part of the change is to "fail fast" in the case where there is no such header in this particular request or response ... so we don't want to call `findHeader(key)` because we already know it isn't there.
> 
> I thought the trace was only useful in the case where the header was present, since the main point of the trace is to produce the header associated with the key. But I can restructure it like this, to avoid the call to `findHeader(key)` and still produce the trace.
> 
> ```
>     public HeaderField getHeader(HeaderKeys key) {
>         if (null == key) {
>             throw new IllegalArgumentException("Null input provided");
>         }
>         HeaderElement elem = null;
>         if (headerKeysByName.get(key.getName()) != null) {
>             elem = findHeader(key);
>         }
>         if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
>             Tr.debug(tc, "getHeader(h): " + key.getName() + " " + elem);
>         }
>         if (null == elem) {
>             return NULL_HEADER;
>         }
> ```

That is the same logic as my suggested change comment.  I just used ? : operations instead of an if statement.